hermes/src/price_feeds_metadata.rs

```diff
@@ -0,0 +1,83 @@
+use {
+    crate::{
+        api::types::{
+            AssetType,
+            PriceFeedMetadata,
+        },
+        network::pythnet::fetch_and_store_price_feeds_metadata,
+        state::{
+            cache::AggregateCache,
+            retrieve_price_feeds_metadata,
+        },
+    },
+    anyhow::Result,
+};
+
+pub async fn get_price_feeds_metadata<S>(
+    state: &S,
+    query: Option<String>,
+    asset_type: Option<AssetType>,
+) -> Result<Vec<PriceFeedMetadata>>
+where
+    S: AggregateCache,
+    S: PriceFeedMetadataProvider,
+{
+    match PriceFeedMetadataProvider::get_price_feeds_metadata(state, query, asset_type).await {
+        Ok(price_feeds_with_update_data) => Ok(price_feeds_with_update_data),
+        Err(e) => Err(e),
+    }
+}
+
+
+#[async_trait::async_trait]
+pub trait PriceFeedMetadataProvider {
+    async fn get_price_feeds_metadata(
+        &self,
+        query: Option<String>,
+        asset_type: Option<AssetType>,
+    ) -> Result<Vec<PriceFeedMetadata>>;
+}
+
+#[async_trait::async_trait]
+impl PriceFeedMetadataProvider for crate::state::State {
+    async fn get_price_feeds_metadata(
+        &self,
+        query: Option<String>,
+        asset_type: Option<AssetType>,
+    ) -> Result<Vec<PriceFeedMetadata>> {
+        let mut price_feeds_metadata = {
+            let feeds = retrieve_price_feeds_metadata(&self).await?;
+            if feeds.is_empty() {
+                // If the result is an empty Vec, fetch and store new price feeds
+                fetch_and_store_price_feeds_metadata(&self).await?
+            } else {
+                feeds
+            }
+        };
+
+
+        // Filter by query if provided
```

Comment:
So this code implies we don't have good structured data to work with. It's better to implement this logic when you're first parsing into `attributes`. You can read more about this idea [here](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/) but what I would suggest is to look at the code where I've commented `Ref: A` and think of this as where we validate.

Instead of reading out the strings, let's strongly type this with:

```
struct Attributes {
    symbol: String,
    asset_type: AssetType,
    ...
}
```

Note that these are not `Option` so we're requiring at read time that our products are compliant. This means later when we get to this code you can replace all the above logic with just:

```
price_feeds_metadata.retain(|feed| feed.symbol.contains(&query_str));
price_feeds_metadata.retain(|feed| feed.asset_type == asset_type);
```

TL;DR we're moving all our validation/parsing logic th the constructor so our logic can be pure logic.

Reply:
this makes sense, will follow up on this with a separate PR